sdk/communication/communication-common/test/public/node/entraCommunicationTokenCredential.spec.ts

```diff
@@ -0,0 +1,130 @@
+      token: acsToken,
+      expiresOn: new Date(Date.now() + 60 * 60 * 1000).toISOString(),
+    },
+  });
+
+describe("Entra CommunicationTokenCredential", function () {
+  afterEach(() => {
+    nock.cleanAll();
+  });
+
+  it("Token exchange is called when passing Entra credential", async function () {
+    const scope = successApiMock();
+    const getTokenSpy: sinon.SinonSpy = sinon.spy(tokenCredential, "getToken");
+
+    const credential = new AzureCommunicationTokenCredential({
+      resourceEndpoint,
+      tokenCredential,
+      scopes,
+    });
+
+    const tokenResult = (await credential.getToken()).token;
+    assert.strictEqual(tokenResult, acsToken);
+    assert.isTrue(getTokenSpy.alwaysCalledWithExactly(scopes, undefined));
+    assert.isTrue(getTokenSpy.callCount > 0);
+    assert.isTrue(scope.isDone());
+    getTokenSpy.restore();
+  });
+
+  it("Retries when service is busy", async function () {
+    const busy = apiMock().reply(
+      503,
+      {
+        error: "Service Unavailable",
+      },
+      {
+        "Retry-After": "0",
+      },
+    );
+
+    const success = successApiMock();
+
+    const credential = new AzureCommunicationTokenCredential({
+      resourceEndpoint,
+      tokenCredential,
+      scopes,
+    });
+
+    const tokenResult = (await credential.getToken()).token;
+    assert.strictEqual(tokenResult, acsToken);
+    assert.isTrue(busy.isDone());
+    assert.isTrue(success.isDone());
+  });
+
+  async function testAbortSignal(abort: boolean) {
+    let scope = successApiMock();
+    const getTokenSpy: sinon.SinonSpy = sinon.spy(tokenCredential, "getToken");
+    const abortController = new AbortController();
+
+    if (abort) {
+      abortController.abort();
+    }
+
+    const credential = new AzureCommunicationTokenCredential({
+      resourceEndpoint,
+      tokenCredential,
+      scopes,
+    });
+
+    // the credential constructor immediately starts fetching,
+    // so await and then reset the internally cached result to force a new token exchange
+    await credential.getToken();
+    assert.isTrue(scope.isDone());
+    const cachedResult = (credential as any).tokenCredential.result;
+    cachedResult.acsToken.token = "";
+    cachedResult.entraToken = "";
+    getTokenSpy.resetHistory();
+    scope = successApiMock();
+
+    // now getToken with abort signal
+    const tokenResult = (await credential.getToken({ abortSignal: abortController.signal })).token;
+
+    if (abort) {
+      assert.strictEqual(tokenResult, "");
+      assert.isFalse(getTokenSpy.called);
+      assert.isFalse(scope.isDone());
+    } else {
+      assert.strictEqual(tokenResult, acsToken);
+      assert.isTrue(getTokenSpy.called);
+      assert.isTrue(scope.isDone());
+    }
+    getTokenSpy.restore();
+  }
+
+  it("Respects abort signal", () => testAbortSignal(true));
+  it("Continues when abort signal isn't aborted", () => testAbortSignal(false));
+
+  // todo: doesn't retry more than 3 times
```

Comment:
There are some more tests that we should now or in a later PR:

* doesn't retry more than three times when service is busy
* when the entra token changes, the exchange endpoint gets called again
* when the entra token stays the same, the exchange endpoint doesn't get called
* when the ACS token has expired, the tokenCredential.getToken gets called and the exchange endpoint gets called
* when the exchange endpoint returns a 4xx, getToken will throw an error